sound/soc/sof/pm.c

```diff
@@ -255,7 +255,7 @@ int snd_sof_resume(struct device *dev)
 	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
 	int ret;
 
-	if (snd_sof_dsp_d0i3_on_suspend(sdev)) {
+	if (sdev->d0i3_suspend_cnt) {
```

Comment:
This direction is not quite right.
It turns out that for SoundWire we do have to keep the DSP powered if we want to avoid complete re-enumeration when exiting clock stop. This could happen in the complete absence of 'D0i3 compatible' streams.
this is really missing an idle loop detection that is based on 
a) D0i3-compatible streams being opened
b) no IPC for a while
c) the state of child devices, e.g. if a SoundWire master remains 'active' preventing pm_runtime entry.

Reply:
@plbossart I didnt quite understand this ask. The change here only really affects system suspend when there are D0i3-compatible streams running in the DSP. Are you saying that we also need to account for the fact that the SDW master could be active?

Reply:
what I am saying is that while doing a system suspend, the device may already be in D0i3 and remain there (S0 target) or move to D3 (S3 target).

The use of refcounts would be quite misleading here, it's only one of the reasons why the device might remain in D0i3.

Reply:
@plbossart agreed wrt the WoV usecase this is the only reason that the DSP would remain in D0i3. Maybe we're missing the line of sight for the other possible indicators for this. Can we possibly think of adding this in a separate PR?

Reply:
well no. You remove a criterion to replace it by a refcount, we are not going to modify this again later to come back to another criterion. We need to respect the time of reviewers...

Reply:
@plbossart I think I know what you concern, I talked with @RanderWang about keeping the DSP in D0I3 for runtime suspend (for SDW clock stop supporting), I did't intend mix that with the D0I3-compatible stream running on D0I3, as I commented in https://github.com/thesofproject/linux/pull/1524#issuecomment-559761166.

Reply:
Hi @plbossart @ranj063 so the state machine could be updated like this:

```
/*
 * Audio DSP states may transform as below:+
 *
 *                +---------------------+  runtime resume
 *                |                     <---------------------+
 *   +------------>                     |  runtime suspend    |
 *   |            |       D0(active)    +----------------+    |
 *   |            |                     +------------+   |    |
 *   |       +----+                     |   D0I3     |   |    |
 *   |       |    +-------^----+----^-- +compatible  |   |    |
 *   |       |            |    |    |      stream    |   |    |
 *   |resume |            |    |    |      opened    |   |    |
 *   | from  |  D3        |    |    |       only     |   |    |
 *   |  D3   |suspend     |    |    |                |   |    |
 *   |       |            |    |    |The last        |   |    |
 *   |       |            |    |    |opened D0i3     |   |    |
 *   |       |            |    |    |compatible      |   |    |
 *   |       |            |    |    |stream closed   |   |    |
 * +-+-------v-------+    |    |    |            +---v---v----+---+
 * |                 |    |    |    +------------+                |
 * |  D3 (suspended) |    |    |                 |      D0I3      |
 * |                 |    |    +----------------->                +---+
 * |                 |    +----------------------+                |   |
 * +-------^---------+         resume from       +----------------+   |
 *         |                   D0I3 suspend                           |
 *         |                                   D3 suspend             |
 *         +----------------------------------------------------------+
 *
 * d0i3_suspend = s0_suspend && D0I3 stream opened,
 * D3 suspend = !d0i3_suspend,
 */


```

Reply:
@keyonjie are you saying that you'd never enter D3 during runtime suspend then>?

Reply:
Yes, @ranj063 enter to D0I3 only for runtime suspend, if the platform doesn't support D0I3, the power gating will still work when it is not used, for power saving.

Reply:
NAK, runtime suspend and D0i3 are mutually incompatible.

Reply:
@plbossart can you help guide how the state machine should look like in your brain?